File: scripts/seed.js
#!/usr/bin/env node
const { MongoClient } = require('mongodb');
const bcrypt = require('bcrypt');
const { randomUUID } = require('crypto');

const CURATED_SYMBOLS = [
  { symbol: 'DANGCEM', name: 'Dangote Cement', sector: 'Industrial Goods' },
  { symbol: 'GTCO', name: 'GTCO Plc', sector: 'Financial Services' },
  { symbol: 'ZENITHBANK', name: 'Zenith Bank', sector: 'Financial Services' },
  { symbol: 'MTNN', name: 'MTN Nigeria', sector: 'ICT' },
  { symbol: 'BUACEMENT', name: 'BUA Cement', sector: 'Industrial Goods' },
  { symbol: 'ACCESSCORP', name: 'Access Holdings', sector: 'Financial Services' },
  { symbol: 'UBA', name: 'UBA', sector: 'Financial Services' },
  { symbol: 'FBNH', name: 'FBN Holdings', sector: 'Financial Services' },
  { symbol: 'SEPLAT', name: 'Seplat Energy', sector: 'Oil & Gas' },
  { symbol: 'NESTLE', name: 'Nestle Nigeria', sector: 'Consumer Goods' },
  { symbol: 'BUAFOODS', name: 'BUA Foods', sector: 'Consumer Goods' },
  { symbol: 'AIRTELAFRI', name: 'Airtel Africa', sector: 'ICT' },
  { symbol: 'WAPCO', name: 'Lafarge Africa', sector: 'Industrial Goods' },
  { symbol: 'GUARANTY', name: 'Guaranty Trust Holding', sector: 'Financial Services' },
  { symbol: 'STANBIC', name: 'Stanbic IBTC', sector: 'Financial Services' },
  { symbol: 'FLOURMILL', name: 'Flour Mills', sector: 'Consumer Goods' },
  { symbol: 'PRESCO', name: 'Presco', sector: 'Agriculture' },
  { symbol: 'OKOMUOIL', name: 'Okomu Oil Palm', sector: 'Agriculture' },
  { symbol: 'NASCON', name: 'Nascon Allied', sector: 'Consumer Goods' },
  { symbol: 'INTBREW', name: 'International Breweries', sector: 'Consumer Goods' },
];

const BASE_PRICES = {
  DANGCEM: 280, GTCO: 45, ZENITHBANK: 38, MTNN: 220, BUACEMENT: 95,
  ACCESSCORP: 22, UBA: 28, FBNH: 18, SEPLAT: 3200, NESTLE: 1200,
  BUAFOODS: 150, AIRTELAFRI: 2100, WAPCO: 35, GUARANTY: 55, STANBIC: 65,
  FLOURMILL: 42, PRESCO: 280, OKOMUOIL: 350, NASCON: 18, INTBREW: 5,
};

function generatePriceHistory(symbol, basePrice, days = 120) {
  const rows = [];
  const today = new Date();
  let price = basePrice;
  for (let i = days; i >= 0; i--) {
    const d = new Date(today);
    d.setDate(d.getDate() - i);
    if (d.getDay() === 0 || d.getDay() === 6) continue;
    const change = (Math.random() - 0.48) * 0.03;
    price = Math.max(price * (1 + change), 1);
    rows.push({
      _id: randomUUID(),
      symbol,
      trade_date: d.toISOString().split('T')[0],
      price: Math.round(price * 100) / 100,
      change_percent: Math.round(change * 10000) / 100,
      volume: Math.floor(Math.random() * 5000000) + 100000,
    });
  }
  return rows;
}

async function seed() {
  const uri = process.env.MONGODB_URI || 'mongodb://localhost:27017/pulsar';
  const client = new MongoClient(uri);
  await client.connect();
  const db = client.db();

  for (const inst of CURATED_SYMBOLS) {
    await db.collection('instruments').updateOne(
      { symbol: inst.symbol },
      {
        $setOnInsert: { _id: randomUUID(), added_at: new Date() },
        $set: { symbol: inst.symbol, name: inst.name, sector: inst.sector, is_active: true },
      },
      { upsert: true },
    );
  }

  for (const inst of CURATED_SYMBOLS) {
    const prices = generatePriceHistory(inst.symbol, BASE_PRICES[inst.symbol] || 100);
    for (const p of prices) {
      await db.collection('price_history').updateOne(
        { symbol: p.symbol, trade_date: p.trade_date },
        { $set: p },
        { upsert: true },
      );
    }
  }

  const today = new Date();
  let asiValue = 95000;
  for (let i = 60; i >= 0; i--) {
    const d = new Date(today);
    d.setDate(d.getDate() - i);
    if (d.getDay() === 0 || d.getDay() === 6) continue;
    asiValue *= 1 + (Math.random() - 0.48) * 0.01;
    const tradeDate = d.toISOString().split('T')[0];
    await db.collection('index_history').updateOne(
      { index_code: 'ASI', trade_date: tradeDate },
      {
        $setOnInsert: { _id: randomUUID() },
        $set: {
          index_code: 'ASI',
          trade_date: tradeDate,
          value: Math.round(asiValue),
          points: Math.round((Math.random() - 0.5) * 200),
        },
      },
      { upsert: true },
    );
  }

  let param = await db.collection('strategy_param_sets').findOne({ name: 'default-sandbox' });
  if (!param) {
    param = {
      _id: randomUUID(),
      name: 'default-sandbox',
      max_position_pct: 0.1,
      max_daily_trades: 5,
      stop_loss_pct: 0.05,
      min_confidence_to_trade: 0.65,
      max_daily_drawdown_pct: 0.03,
      allowed_symbols: null,
      position_size_pct: 0.05,
      is_active: true,
      created_at: new Date(),
    };
    await db.collection('strategy_param_sets').insertOne(param);
  } else {
    await db.collection('strategy_param_sets').updateOne(
      { _id: param._id },
      { $set: { allowed_symbols: null, is_active: true } },
    );
  }

  await db.collection('strategy_param_sets').updateMany(
    { _id: { $ne: param._id } },
    { $set: { is_active: false } },
  );

  const startingCapital = Number(process.env.DEFAULT_STARTING_CAPITAL || 10000000);
  await db.collection('sandbox_portfolios').updateOne(
    { name: 'default-sandbox' },
    {
      $setOnInsert: { _id: randomUUID(), created_at: new Date() },
      $set: {
        name: 'default-sandbox',
        starting_capital: startingCapital,
        cash_balance: startingCapital,
        strategy_param_set_id: param._id,
      },
    },
    { upsert: true },
  );

  const passwordHash = await bcrypt.hash('admin123', 10);
  await db.collection('users').updateOne(
    { email: 'admin@ngx.local' },
    {
      $setOnInsert: { _id: randomUUID(), created_at: new Date() },
      $set: { email: 'admin@ngx.local', password_hash: passwordHash },
    },
    { upsert: true },
  );

  await client.close();
  console.log('MongoDB seed complete.');
}

seed().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: apps/desktop/src-tauri/src/seed.rs
use anyhow::Result;
use rusqlite::Connection;
use uuid::Uuid;

pub struct SeedService;

const CURATED: &[(&str, &str, &str)] = &[
    ("DANGCEM", "Dangote Cement", "Industrial Goods"),
    ("GTCO", "GTCO Plc", "Financial Services"),
    ("ZENITHBANK", "Zenith Bank", "Financial Services"),
    ("MTNN", "MTN Nigeria", "ICT"),
    ("BUACEMENT", "BUA Cement", "Industrial Goods"),
    ("ACCESSCORP", "Access Holdings", "Financial Services"),
    ("UBA", "UBA", "Financial Services"),
    ("FBNH", "FBN Holdings", "Financial Services"),
    ("SEPLAT", "Seplat Energy", "Oil & Gas"),
    ("NESTLE", "Nestle Nigeria", "Consumer Goods"),
    ("BUAFOODS", "BUA Foods", "Consumer Goods"),
    ("AIRTELAFRI", "Airtel Africa", "ICT"),
    ("WAPCO", "Lafarge Africa", "Industrial Goods"),
    ("GUARANTY", "Guaranty Trust Holding", "Financial Services"),
    ("STANBIC", "Stanbic IBTC", "Financial Services"),
    ("FLOURMILL", "Flour Mills", "Consumer Goods"),
    ("PRESCO", "Presco", "Agriculture"),
    ("OKOMUOIL", "Okomu Oil Palm", "Agriculture"),
    ("NASCON", "Nascon Allied", "Consumer Goods"),
    ("INTBREW", "International Breweries", "Consumer Goods"),
];

const BASE_PRICES: &[(&str, f64)] = &[
    ("DANGCEM", 280.0), ("GTCO", 45.0), ("ZENITHBANK", 38.0), ("MTNN", 220.0), ("BUACEMENT", 95.0),
    ("ACCESSCORP", 22.0), ("UBA", 28.0), ("FBNH", 18.0), ("SEPLAT", 3200.0), ("NESTLE", 1200.0),
    ("BUAFOODS", 150.0), ("AIRTELAFRI", 2100.0), ("WAPCO", 35.0), ("GUARANTY", 55.0), ("STANBIC", 65.0),
    ("FLOURMILL", 42.0), ("PRESCO", 280.0), ("OKOMUOIL", 350.0), ("NASCON", 18.0), ("INTBREW", 5.0),
];

pub const CRYPTO_USDT_STARTING: f64 = 10_000.0;

pub const CRYPTO_UNIVERSE: &[(&str, &str)] = &[
    ("BTCUSDT", "Bitcoin"),
    ("ETHUSDT", "Ethereum"),
    ("SOLUSDT", "Solana"),
    ("BNBUSDT", "BNB"),
    ("XRPUSDT", "XRP"),
    ("DOGEUSDT", "Dogecoin"),
    ("ADAUSDT", "Cardano"),
    ("AVAXUSDT", "Avalanche"),
    ("LINKUSDT", "Chainlink"),
    ("DOTUSDT", "Polkadot"),
    ("POLUSDT", "Polygon"),
    ("LTCUSDT", "Litecoin"),
    ("ATOMUSDT", "Cosmos"),
    ("UNIUSDT", "Uniswap"),
    ("NEARUSDT", "NEAR"),
    ("FILUSDT", "Filecoin"),
    ("ARBUSDT", "Arbitrum"),
    ("OPUSDT", "Optimism"),
];

impl SeedService {
    pub fn seed_if_empty(conn: &Connection, starting_capital: f64) -> Result<()> {
        let count: i64 = conn.query_row("SELECT COUNT(*) FROM instruments", [], |row| row.get(0))?;
        if count > 0 {
            return Ok(());
        }
        Self::seed_all(conn, starting_capital)
    }

    pub fn seed_all(conn: &Connection, starting_capital: f64) -> Result<()> {
        for (symbol, name, sector) in CURATED {
            conn.execute(
                "INSERT INTO instruments (symbol, name, sector, is_active, module) VALUES (?1, ?2, ?3, 1, 'stocks')",
                rusqlite::params![symbol, name, sector],
            )?;
        }

        for (symbol, name, sector) in CURATED {
            let base = BASE_PRICES.iter().find(|(s, _)| *s == *symbol).map(|(_, p)| *p).unwrap_or(100.0);
            let prices = generate_price_history(symbol, base, 120);
            for p in prices {
                conn.execute(
                    "INSERT OR IGNORE INTO price_history (symbol, trade_date, price, change_percent, volume)
                     VALUES (?1, ?2, ?3, ?4, ?5)",
                    rusqlite::params![p.0, p.1, p.2, p.3, p.4],
                )?;
            }
            let _ = (name, sector);
        }

        let param_id = Uuid::new_v4().to_string();
        conn.execute(
            "INSERT INTO strategy_param_sets (id, name, is_active, allowed_symbols)
             VALUES (?1, 'default', 1, NULL)",
            rusqlite::params![param_id],
        )?;

        let portfolio_id = Uuid::new_v4().to_string();
        conn.execute(
            "INSERT INTO sandbox_portfolios (id, name, starting_capital, cash_balance, strategy_param_set_id)
             VALUES (?1, 'default-sandbox', ?2, ?2, ?3)",
            rusqlite::params![portfolio_id, starting_capital, param_id],
        )?;

        Ok(())
    }

    pub fn ensure_crypto_sandbox(conn: &Connection) -> Result<()> {
        for (symbol, name) in CRYPTO_UNIVERSE {
            conn.execute(
                "INSERT INTO instruments (symbol, name, sector, is_active, module)
                 VALUES (?1, ?2, 'Crypto', 1, 'crypto')
                 ON CONFLICT(symbol) DO UPDATE SET
                   name = excluded.name,
                   sector = 'Crypto',
                   module = 'crypto'",
                rusqlite::params![symbol, name],
            )?;
        }

        let exists: i64 = conn.query_row(
            "SELECT COUNT(*) FROM sandbox_portfolios WHERE name = 'default-crypto-sandbox'",
            [],
            |r| r.get(0),
        )?;
        if exists > 0 {
            return Ok(());
        }
        let param_id: String = conn.query_row(
            "SELECT id FROM strategy_param_sets WHERE is_active = 1 LIMIT 1",
            [],
            |r| r.get(0),
        )?;
        let portfolio_id = Uuid::new_v4().to_string();
        conn.execute(
            "INSERT INTO sandbox_portfolios (id, name, starting_capital, cash_balance, strategy_param_set_id)
             VALUES (?1, 'default-crypto-sandbox', ?2, ?2, ?3)",
            rusqlite::params![portfolio_id, CRYPTO_USDT_STARTING, param_id],
        )?;
        Ok(())
    }
}

fn generate_price_history(symbol: &str, base_price: f64, days: i64) -> Vec<(String, String, f64, f64, i64)> {
    let _ = symbol;
    let mut rows = vec![];
    let today = chrono::Utc::now();
    let mut price = base_price;
    for i in (0..=days).rev() {
        let d = today - chrono::Duration::days(i);
        if d.weekday().num_days_from_monday() >= 5 {
            continue;
        }
        let change = (rand_simple(i) - 0.48) * 0.03;
        price = (price * (1.0 + change)).max(1.0);
        rows.push((
            symbol.to_string(),
            d.format("%Y-%m-%d").to_string(),
            (price * 100.0).round() / 100.0,
            (change * 10000.0).round() / 100.0,
            100_000 + (i * 1000) as i64,
        ));
    }
    rows
}

fn rand_simple(seed: i64) -> f64 {
    let x = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7fffffff;
    (x as f64) / 0x7fffffff as f64
}

use chrono::Datelike;
